src/lib/x509/pkix_enums.h

```diff
@@ -100,18 +102,76 @@ BOTAN_PUBLIC_API(2,0) const char* to_string(Certificate_Status_Code code);
 * X.509v3 Key Constraints.
 * If updating update copy in ffi.h
 */
-enum Key_Constraints {
-   NO_CONSTRAINTS     = 0,
-   DIGITAL_SIGNATURE  = 1 << 15,
-   NON_REPUDIATION    = 1 << 14,
-   KEY_ENCIPHERMENT   = 1 << 13,
-   DATA_ENCIPHERMENT  = 1 << 12,
-   KEY_AGREEMENT      = 1 << 11,
-   KEY_CERT_SIGN      = 1 << 10,
-   CRL_SIGN           = 1 << 9,
-   ENCIPHER_ONLY      = 1 << 8,
-   DECIPHER_ONLY      = 1 << 7
-};
+class BOTAN_PUBLIC_API(3,0) Key_Constraints
+   {
+   public:
+      enum Key_Constraints_Bits : uint32_t {
+         None              = 0,
+         DigitalSignature  = 1 << 15,
+         NonRepudiation    = 1 << 14,
+         KeyEncipherment   = 1 << 13,
+         DataEncipherment  = 1 << 12,
+         KeyAgreement      = 1 << 11,
+         KeyCertSign       = 1 << 10,
+         CrlSign           = 1 << 9,
+         EncipherOnly      = 1 << 8,
+         DecipherOnly      = 1 << 7,
+
+         // Deprecated SHOUTING_CASE names for Key_Constraints
+         // will be removed in a future major release
+         NO_CONSTRAINTS    = None,
+         DIGITAL_SIGNATURE = DigitalSignature,
+         NON_REPUDIATION   = NonRepudiation,
+         KEY_ENCIPHERMENT  = KeyEncipherment,
+         DATA_ENCIPHERMENT = DataEncipherment,
+         KEY_AGREEMENT     = KeyAgreement,
+         KEY_CERT_SIGN     = KeyCertSign,
+         CRL_SIGN          = CrlSign,
+         ENCIPHER_ONLY     = EncipherOnly,
+         DECIPHER_ONLY     = DecipherOnly,
+      };
```

Comment:
Let's split this up into two enums (new and deprecated). This might increase the number of required overloads in the methods and operators below, but allow us to mark the old stuff as deprecated.

Here's a sketch in Godbolt: https://godbolt.org/z/z81r4oGrz

Reply:
Nice :+1:

Reply:
I discovered there is a way that is (IMO) cleaner:

```
enum E {
  Something = 0,

  SOME_LOUD_THING [[deprecated("Use Something")]] = Something,
}
```

warns in exactly the way we would want, and without having to create a second type.